File: content/Storage.ts
// Copyright (c) 2018 cloudcrate solutions UG (haftungsbeschraenkt)

const storageAssembly = 'Cloudcrate_AspNetCore_Blazor_Browser';
const storageNamespace = `${storageAssembly}_Storage`;

const storages: { [key: string]: Storage } = {
    LocalStorage: localStorage,
    SessionStorage: sessionStorage
};

for (var storageTypeName in storages) {
    if (storages.hasOwnProperty(storageTypeName)) {
        const storage = storages[storageTypeName];
        const storageFullTypeName = `${storageNamespace}_${storageTypeName}`;

        window[storageFullTypeName] = {
            Clear: () => {
                clear(storage);
            },

            GetItem: (key: string) => {
                return getItem(storage, key);
            },

            Key: (index: number) => {
                return key(storage, index);
            },

            Length: () => {
                return getLength(storage);
            },

            RemoveItem: (key: string) => {
                removeItem(storage, key);
            },

            SetItem: (key: string, data: any) => {
                setItem(storage, key, data);
            },

            GetItemString: (key: string) => {
                return getItemString(storage, key);
            },

            SetItemString: (key: string, data: string) => {
                setItemString(storage, key, data);
            },

            GetItemNumber: (index: number) => {
                return getItemNumber(storage, index);
            },

            SetItemNumber: (index: number, data: string) => {
                setItemNumber(storage, index, data);
            }

        };
    }
}

function clear(storage: Storage) {
    storage.clear();
}

function getItem(storage: Storage, key: string) {
    return storage.getItem(key);
}

function key(storage: Storage, index: number) {
    return storage.key(index);
}

function getLength(storage: Storage) {
    return storage.length;
}

function removeItem(storage: Storage, key: string) {
    storage.removeItem(key);
}

function setItem(storage: Storage, key: string, data: any) {
    storage.setItem(key, data);
}

function getItemString(storage: Storage, key: string) {
    return storage[key];
}

function setItemString(storage: Storage, key: string, data: any) {
    storage[key] = data;
}

function getItemNumber(storage: Storage, index: number) {
    return storage[index];
}

function setItemNumber(storage: Storage, index: number, data: string) {
    storage[index] = data;
}

File: Storage.cs
// Copyright (c) 2018 cloudcrate solutions UG (haftungsbeschraenkt)

using System;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.DependencyInjection.Extensions;
using Microsoft.JSInterop;

namespace Cloudcrate.AspNetCore.Blazor.Browser.Storage
{
    public abstract class StorageBase
    {
        private readonly IJSRuntime _jsRuntime;
        private readonly IJSInProcessRuntime _jsProcessRuntime;
        private readonly string _fullTypeName;

        private EventHandler<StorageEventArgs> _storageChanged;

        protected abstract string StorageTypeName { get; }


        protected internal StorageBase(IJSRuntime jsRuntime)
        {
            _jsRuntime = jsRuntime;
            _jsProcessRuntime = (IJSInProcessRuntime)_jsRuntime;
            _fullTypeName = GetType().FullName.Replace('.', '_');
        }

        public void Clear()
        {
            _jsProcessRuntime.Invoke<object>($"{_fullTypeName}.Clear");
        }

        public string GetItem(string key)
        {
            return _jsProcessRuntime.Invoke<string>($"{_fullTypeName}.GetItem", key);
        }

        public T GetItem<T>(string key)
        {
            var json = GetItem(key);
            return string.IsNullOrEmpty(json) ? default(T) : Json.Deserialize<T>(json);
        }

        public string Key(int index)
        {
            return _jsProcessRuntime.Invoke<string>($"{_fullTypeName}.Key", index);
        }

        public int Length => _jsProcessRuntime.Invoke<int>($"{_fullTypeName}.Length");

        public void RemoveItem(string key)
        {
            _jsProcessRuntime.Invoke<object>($"{_fullTypeName}.RemoveItem", key);
        }

        public void SetItem(string key, string data)
        {
            _jsProcessRuntime.Invoke<object>($"{_fullTypeName}.SetItem", key, data);
        }

        public void SetItem(string key, object data)
        {
            SetItem(key, Json.Serialize(data));
        }

        public string this[string key]
        {
            get => _jsProcessRuntime.Invoke<string>($"{_fullTypeName}.GetItemString", key);
            set => _jsProcessRuntime.Invoke<object>($"{_fullTypeName}.SetItemString", key, value);
        }

        public string this[int index]
        {
            get => _jsProcessRuntime.Invoke<string>($"{_fullTypeName}.GetItemNumber", index);
            set => _jsProcessRuntime.Invoke<object>($"{_fullTypeName}.SetItemNumber", index, value);
        }

        public event EventHandler<StorageEventArgs> StorageChanged
        {
            add
            {
                if (_storageChanged == null)
                {
                    this._jsRuntime.InvokeAsync<object>(
                        $"{_fullTypeName}.AddEventListener",
                        new DotNetObjectRef(this)
                    );
                }
                _storageChanged += value;
            }
            remove
            {
                _storageChanged -= value;
                if (_storageChanged == null)
                {
                    this._jsRuntime.InvokeAsync<object>($"{_fullTypeName}.RemoveEventListener");
                }
            }
        }

        [JSInvokable]
        public virtual void OnStorageChanged(string key, object oldValue, object newValue)
        {
            EventHandler<StorageEventArgs> handler = _storageChanged;
            if (handler != null)
            {
                handler(this, new StorageEventArgs
                {
                    Key = key,
                    OldValue = oldValue,
                    NewValue = newValue,
                });
            }
        }
    }

    public sealed class LocalStorage : StorageBase
    {
        protected override string StorageTypeName => nameof(LocalStorage);

        public LocalStorage(IJSRuntime jsRuntime) : base(jsRuntime)
        {
        }
    }

    public sealed class SessionStorage : StorageBase
    {
        protected override string StorageTypeName => nameof(SessionStorage);

        public SessionStorage(IJSRuntime jsRuntime) : base(jsRuntime)
        {
        }
    }

    public static class ServiceCollectionExtensions
    {
        public static void AddStorage(this IServiceCollection col)
        {
            col.TryAddSingleton<LocalStorage>();
            col.TryAddSingleton<SessionStorage>();
        }
    }
}
